src/cpp/session/SessionSourceDatabase.cpp

```diff
@@ -472,7 +474,13 @@ void SourceDocument::writeToJson(json::Object* pDocJson) const
    jsonDoc["source_window"] = sourceWindow_;
    jsonDoc["last_content_update"] = json::Value(
          static_cast<boost::int64_t>(lastContentUpdate_));
+}
 
+SEXP SourceDocument::toRObject(r::sexp::Protect* pProtect) const
+{
+   json::Object object;
+   writeToJson(&object);
```

Comment:
FYI this can be expensive for larger documents -- `writeToJson` actually dumps the entire contents of the file into JSON (`contents` is one of the properties).